app/validators/list.ts

```diff
@@ -0,0 +1,49 @@
+import { ValidatorFunction, ValidatorResult } from 'ember-changeset-validations';
+
+export type ListValidatorFunction = (
+    key: string,
+    newValue: undefined | unknown,
+    oldValue: undefined | unknown,
+    changes: Record<string, unknown>,
+    content: Record<string, unknown>,
+) => true | ValidatorResult[];
+
+/**
+ * Validate each item in a list using the given validator.
+ */
+export default function validateList(validator: ValidatorFunction): ListValidatorFunction {
+    return (key, newValue, oldValue, changes, content) => {
+        if (typeof newValue === 'undefined') {
+            // This validator validates list elements.
+            // If there is no list, there's nothing to validate!
+            return true;
+        }
+        if (!Array.isArray(newValue) || (typeof oldValue !== 'undefined' && !Array.isArray(oldValue))) {
+            throw new Error('validateList can only be used for array properties');
+        }
+        const results: ValidatorResult[] = [];
+        newValue.forEach((newItem, index) => {
+            const result = validator(
+                key,
+                newItem,
+                oldValue ? oldValue[index] : undefined,
+                changes,
+                content,
+            );
+            results.push(result);
+        });
+        return results.every(itemResult => itemResult === true) ? true : results;
+    };
+}
+
+/**
+ * Transpose a ValidatorResult table so that items are the first dimension and
+ * validator results for each item are the second dimension, excluding validator
+ * results that are not arrays (and thus aren't for individual list items).
+ */
+export function transposeResults(results?: Array<ValidatorResult | ValidatorResult[]>) {
+    const transposer = (acc: ValidatorResult[][], itemErrors: ValidatorResult[]) => (
+        Array.isArray(itemErrors) ? itemErrors.map((val, index) => [...(acc[index] || []), val]) : acc
```

Comment:
Could be further simplified with a `.filter(Array.isArray)` -- anytime the body of a `reduce`/`forEach`/etc is `if`-wrapped, `filter` is nice.

Reply:
Lol, yep, that's def better.